File: MS.NET/Fundamentals/Generic/Basics/genclasstest.cs
using System;

class SimpleStack<V>
{
	class Node
	{
		internal V Value;
		internal Node Below;

		internal Node(V value, Node below)
		{
			Value = value;
			Below = below;
		}
	}

	private Node top;

	public void Push(V value)	
	{
		top = new Node(value, top);
	}

	public V Pop()
	{
		V value = top.Value;
		top = top.Below;

		return value;
	}

	public bool Empty()
	{
		return top == null;
	}
}

static class Program
{
	public static void Main()
	{
		SimpleStack<string> a = new SimpleStack<string>();
		a.Push("Monday");
		a.Push("Tuesday");
		a.Push("Wednesday");
		a.Push("Thursday");
		a.Push("Friday");

		SimpleStack<string> b = new SimpleStack<string>();
		b.Push("June");
		b.Push("May");
		b.Push("April");
		b.Push("March");

		var c = new SimpleStack<Interval>();
		c.Push(new Interval(6, 51));
		c.Push(new Interval(7, 12));
		c.Push(new Interval(4, 23));
		c.Push(new Interval(5, 34));
		c.Push(new Interval(3, 45));

		var d = new SimpleStack<object>();
		d.Push("Saturday");
		d.Push(new Interval(2, 30));
		d.Push(12.3);

		while(!a.Empty())
			Console.WriteLine(a.Pop());
		Console.WriteLine("-----------------------------");
		while(!b.Empty())
			Console.WriteLine(b.Pop());
		Console.WriteLine("-----------------------------");
		while(!c.Empty())
			Console.WriteLine(c.Pop());
		Console.WriteLine("-----------------------------");
		while(!d.Empty())
			Console.WriteLine(d.Pop());
	}
}

File: Java/Fundamentals/Generics/Basics/GenClassTest.java
class GenClassTest{

	static class SimpleStack<V>{
		
		class Node{
			
			V value;
			Node below;
			
			Node(V val){
				value = val;
				below = top;
			}
		}

		private Node top;

		public void push(V value){
			top = new Node(value);
		}

		public V pop(){
			V result = top.value;
			top = top.below;
			return result;
		}

		public boolean empty(){
			return top == null;
		}
	}

	public static void main(String[] args){
		SimpleStack<String> a = new SimpleStack<String>();
		a.push("Monday");
		a.push("Tuesday");
		a.push("Wednesday");
		a.push("Thursday");
		a.push("Friday");
		//a.push(12.3);
		SimpleStack<String> b = new SimpleStack<>();
		b.push("June");
		b.push("May");
		b.push("April");
		b.push("March");
		SimpleStack<Interval> c = new SimpleStack<>();
		c.push(new Interval(4, 31));
		c.push(new Interval(7, 12));
		c.push(new Interval(5, 23));
		c.push(new Interval(6, 54));
		c.push(new Interval(3, 45));
		SimpleStack<Double> d = new SimpleStack<>();
		d.push(23.4);
		d.push(34.5);
		d.push(45.6);
		while(!a.empty())
			System.out.println(a.pop());
		System.out.println("-------------------------");
		while(!b.empty())
			System.out.println(b.pop());
		System.out.println("-------------------------");
		while(!c.empty())
			System.out.println(c.pop());
		System.out.println("-------------------------");
		while(!d.empty())
			System.out.println(d.pop());
	}
}


